### PR TITLE
Power Distribution Panel

### PR DESCRIPTION
**Pull request by NickVerwey**
This is my attempt to merge my branch with the master branch, committing the changes I've been working on.

### CODE REVIEW

File: src/decoder/power_distribution_panel.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    DecodeStatus1(data, timestamp)
+        Runs GetCurrent() for the relevent data and updates the status1 
+        dictionary
+    DecodeStatus2(data, timestamp)
+        Runs GetCurrent() for the relevent data and updates the status2
+        dictionary
+    DecodeStatus3(data, timestamp)
+        Runs GetCurrent() as well as GetVoltage() for the relevent data and
+        updates the status3 dictionary 
+    GetCurrent(data, arbid)
```
Comment (ejmccalla) [inline src/decoder/power_distribution_panel.py]: Need to remove the arbid function input.

File: src/decoder/power_distribution_panel.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        
+        self.status1 = {
+            'timestamp' : 0.0,
+            'chan_0' : 0.0,
+            'chan_1' : 0.0,
+            'chan_2' : 0.0,
+            'chan_3' : 0.0,
+            'chan_4' : 0.0,
+            'chan_5' : 0.0
+
```
Comment (ejmccalla) [inline src/decoder/power_distribution_panel.py]: Remove space

File: src/decoder/power_distribution_panel.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            'timestamp'  : 0.0,
+            'chan_6' : 0.0,
+            'chan_7' : 0.0,
+            'chan_8' : 0.0,
+            'chan_9' : 0.0,
+            'chan_10' : 0.0,
+            'chan_11' : 0.0
+        }
+        self.status3 = {
+            'timestamp'  : 0.0,
```
Comment (ejmccalla) [inline src/decoder/power_distribution_panel.py]: spacing on this line and

Comment (ejmccalla) [reply src/decoder/power_distribution_panel.py]: line 84

File: src/decoder/power_distribution_panel.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        data : int
+            An integer input of the 8-bytes of data
+        timestamp : float
+            The time the CAN controller received the data frame
+        """
+        self.status1['timestamp'] = timestamp
+        CurrentList = self.GetCurrent(data)
+        keys = ['chan_0', 'chan_1','chan_2','chan_3','chan_4','chan_5']
+        self.status1.update(dict(zip(keys, CurrentList)))
+
```
[the remark this answers is not in the document]
Comment (NickVerwey) [reply src/decoder/power_distribution_panel.py]: That's what I was trying to find syntax for but couldn't really phrase my search correctly. I'll update it

File: src/decoder/power_distribution_panel.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        
+        Parameters
+        ----------
+        arbid : int
+            The base arbitration ID for each Status frame
+        data : int
+            An integer input of the 8-bytes of data
+        """        
+        CurrentScalar = 0.125
+        CurrentList = [0.0,0.0,0.0,0.0,0.0,0.0]
```
Comment (ejmccalla) [inline src/decoder/power_distribution_panel.py]: For the bitwise operators below, these should be ints and not floats

File: src/decoder/power_distribution_panel.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        
+        hexdata = hex(data)
+        
+        CurrentList[0] = int(str(hexdata)[0:2] + str(hexdata)[-2:],16)
+        CurrentList[0] <<= 2
+        CurrentList[0] |= ((data >> 14) & 0x03)
+
```
Comment (ejmccalla) [inline src/decoder/power_distribution_panel.py]: Not sure about what line 156 is doing.  Based on their code, you just want the bottom 8 bits of the 64 bit _cache.  They achieve this by typing the data to byte.  We can do it by something like this:
CurrentList[0] = data && 0xFF


Their code:
_cache_words[0] = (Int16)((byte)(_cache));
_cache_words[0] <<= 2;
_cache_words[0] |= (Int16)((_cache >> 14) & 0x03);

File: src/decoder/power_distribution_panel.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        ----------
+        arbid : int
+            The base arbitration ID for each Status frame
+        data : int
+            An integer input of the 8-bytes of data
+        """        
+        CurrentScalar = 0.125
+        CurrentList = [0.0,0.0,0.0,0.0,0.0,0.0]
+        
+        hexdata = hex(data)
```
Comment (ejmccalla) [inline src/decoder/power_distribution_panel.py]: you really don't need to convert to hex

File: src/decoder/power_distribution_panel.py (hunk 6 above)
Comment (NickVerwey) [reply src/decoder/power_distribution_panel.py]: Alright, I'll fix that. This was a hack and slash solution that worked by converting the data to a string in hex and removing all portions other than the 0x and the last 2 hex characters (0x0408 would become 0x08), then changing the string back to an integer

File: src/decoder/power_distribution_panel.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        the result to the DecodeStatus3 method
+        
+        Parameters
+        ----------
+        data : int
+            An integer input of the 8-bytes of data
+        """          
+        VoltNum = hex(data >> 48)
+        VoltNum = int(str(VoltNum)[0:2] + str(VoltNum)[-2:],16)
+
```
Comment (ejmccalla) [inline src/decoder/power_distribution_panel.py]: don't need the hex function.  Also, slicing up strings isn't as efficient as doing bitwise operators.  Grab the byte 6 like this:
VoltNum = (data >> 48) && 0xFF

            /* grab the vbat byte */
            byte vbatByte = (byte)(_cache >> 48); /* byte 6 */
            /* scale it to volts */
            float retval = 0.05f * vbatByte + 4.0f;
            /* error handle and return voltage to caller */
            HandleError(errCode);
            return retval;

File: src/decoder/power_distribution_panel.py (hunk 3 above)
Comment (ejmccalla) [reply src/decoder/power_distribution_panel.py]: spacing on line 83

File: src/decoder/power_distribution_panel.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        ----------
+        data : int
+            An integer input of the 8-bytes of data
+        """          
+        VoltNum = ((data >> 48) & 0xFF)
+        retval = (.05 * VoltNum) + 4
+        return retval
+    
+    
+"""
```
Comment (ejmccalla) [inline src/decoder/power_distribution_panel.py]: We will need to move the testing into the test folder.  For now, just remove the code.

**Pull request merged**